docs/adrs/003-bugzilla-response-codes.md

```diff
@@ -0,0 +1,185 @@
+# Bugzilla disables webhooks after too many errors
+
+- Status: proposed
+- Date: 2023-11-27
+
+Tracking issues: 
+- [21 - Setting up Webhooks with 'ANY' Product](https://github.com/mozilla/jira-bugzilla-integration/issues/21)
+- [82 - Exception in JBI can delay webhook delivery](https://github.com/mozilla/jira-bugzilla-integration/issues/82)
+- [181 - Can we ping Bugzilla to see if the webhook is enabled?](https://github.com/mozilla/jira-bugzilla-integration/issues/181)
+- [710 - System is able to "drop" erring messages without user interference](https://github.com/mozilla/jira-bugzilla-integration/issues/710)
+- [730 - Establish convention for capturing system incidents](https://github.com/mozilla/jira-bugzilla-integration/issues/730)
+- [743 - Create alerts for when bugs fail to sync](https://github.com/mozilla/jira-bugzilla-integration/issues/743)
+
+## Context and Problem Statement
+When Bugzilla receives too many error responses from JBI, it stops triggering webhook calls for the entire project causing data to stop syncing. Frequently, these errors are due to a configuration error in Jira or Bugzilla. JBI being unable to process a payload due to errors in configuration (or incomplete configuration) in Jira or a mismatch of data for a single bug. These outages can last multiple days in some cases.
+
+We don't want the entire sync process to stop because of this. We have identified four options to solve this problem.
+
+## Decision Drivers
+- Amount of initial engineering effort
+- Amount of maintenance effort
+- Overall performance of JBI (how quickly is data able to move)
+- How intuitive the solution is to the users that depend on the data (will picking the easiest option solve their needs?)
+
+## Proposed Solution
+We propose to use a data bucket as DLQ. Events can attempt to be reprocessed from here, but we will skip any events identified as stale. See the diagram below:
+
+![Flow chart detailing the data flow, see expandable below for full details](./003.drawio.jpg "Proposed Solution Flow Chart")
+
+<details>
+
+  <summary>Breakdown of flowchart</summary>
+
+  1. JBI receives a payload from Bugzilla or the Retry Scheduler.
+  1. JBI will always return 200/OK for a response.
+  1. If the bug is private, discard the event and log why.
+  1. If the bug cannot be found in the Bugzilla API, discard the event and log why.
+  1. If an associated action cannot be found for the event, discard the event and log why.
+  1. If a matching Jira issue cannot be found, and the event is not creating one, discard the event and log why.
+  1. If there is a mismatch between project keys in the event and Jira, discard the event and log why.
+  1. If there is a blocking event in the DLQ, skip to the Error Event Handler.
+  1. If there is a merge-able event in the DLQ, merge with the current event (current event wins conflicts).
+  1. Write updated data to Jira's API.
+      1. If successful, delete any associated items in DLQ.
```

Comment:
Don't we always want to delete the items from DLQ? I mean, if it fails again, it will be added again no?

Reply:
Process ordering becomes important here, so I wouldn't want to delete a potentially blocking event before attempting to reprocess it just in case we receive another event around the same time.

Also, touching the same file with multiple operations quickly in a data bucket can be messy. GCP has a rate limit of 1 request per second per object. Sending a delete request followed by a write request shortly after might cause issues if we do it too quickly.